sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -109,16 +110,33 @@ public void processElement(
   public static class ViaRandomKey<T> extends PTransform<PCollection<T>, PCollection<T>> {
     private ViaRandomKey() {}
 
+    private ViaRandomKey(@Nullable Integer numBuckets) {
+      this.numBuckets = numBuckets;
+    }
+
+    // The number of buckets to shard into. This is a performance optimization to prevent having
+    // unit sized bundles on the output. If unset, uses a random integer key.
+    private @Nullable Integer numBuckets;
+
+    public ViaRandomKey<T> withNumBuckets(@Nullable Integer numBuckets) {
```

Comment:
Did you run into an issue when using existing Reshuffle implementation ?

Reply:
It was suggested by millsd@ above in this pr that it would have degenerate performance and lead to single key bundles, so he suggested I rewrite it like this.